### PR TITLE
fix memtracker bugs during stress test on graphd and storaged

### PR DESCRIPTION
**Pull request by codesigner**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

##### graphd
- fix multiple not throw safe code path
   - fix Value's setXXX, should set type only when set ptr succedd, during request deserialize;
   - fix ObjectPool deadlock if throw on alloc memory
   - fix Symbol & ExecutionContext's valueMap_, not thread-safe when multiple Executor run concurrently

- fix scheduler
   - fix AsyncMsgNotifyBasedScheduler's scheduling logic, add wait mechinism to guarantee Executors all finish

  

##### storaged
- write operators:  turn off memtracker on stroage's write path
- read operators:  fix uncaught execption & memory leak for storage's read Operators

   
##### others
   - remove noexcept of functions those need alloc memory;
   - add MemoryCheckOffGuard to MetaClient's functions, meta need little memory, but they are important, turn off check on all these functions.

##### Tests
- graphd
   - running days on limited memory, triger memory exceeded for most queries, no crashed;
- storaged
  - run through three test scenario, not crashed,  in https://github.com/vesoft-inc/nebula-ent/issues/2221
     - python3 run.py stress run -scenario go.Go3StepEdge
     - python3 run.py stress run -scenario lookup.LookUp
     - python3 run.py stress run -scenario  find_path.FindShortestPath

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Review by critical27 — APPROVED** [review]
Well done~ Hope we could survive long term test...

**Review by xtcyclist — DISMISSED** [review]

**Review by critical27 — APPROVED** [review]

File: src/common/memory/MemoryUtils.cpp [hunk 1]
```diff
@@ -118,18 +127,77 @@ StatusOr<bool> MemoryUtils::hitsHighWatermark() {
...
+      if (trackable > 0) {
+        MemoryStats::instance().setLimit(trackable * limitRatio);
+      } else {
+        LOG(ERROR) << "Total memory less than untracked " << untrackedMb << " Mib,"
+                   << " MemoryTracker limit is not set properly, "
+                   << " Current memory stats:" << MemoryStats::instance().toString();
+      }
+      LOG(INFO) << "MemoryTracker set static ratio: " << limitRatio;
+    }
+  } else if (limitRatio == 2.0) {
```
Comment (yixinglu) [inline src/common/memory/MemoryUtils.cpp:173]: best not to use equals to compare floating point data types

File: src/common/memory/MemoryUtils.cpp [hunk 2]
```diff
@@ -118,18 +127,77 @@ StatusOr<bool> MemoryUtils::hitsHighWatermark() {
...
+     *           limit = current_used_memory + (available - untracked) * availableRatio
+     */
+    double trackableAvailable = available - (untrackedMb * MiB);
+    if (trackableAvailable > 0) {
+      MemoryStats::instance().updateLimit(trackableAvailable * availableRatio);
+    } else {
+      MemoryStats::instance().updateLimit(0);
+    }
+    DLOG(INFO) << "MemoryTracker set to dynamic self adaptive";
+  } else if (limitRatio == 3.0) {
```
Comment (yixinglu) [inline src/common/memory/MemoryUtils.cpp:185]: ditto

File: src/common/memory/MemoryUtils.cpp [hunk 3]
```diff
@@ -118,18 +127,77 @@ StatusOr<bool> MemoryUtils::hitsHighWatermark() {
...
+    /**
+     * (== 3.0): Special value for disable memory_tracker;
+     *           limit is set to max
+     */
+    if (kCurrentLimitRatio_ != limitRatio) {
```
Comment (yixinglu) [inline src/common/memory/MemoryUtils.cpp:190]: ditto

File: src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp [hunk 4]
```diff
@@ -225,17 +239,94 @@ void AsyncMsgNotifyBasedScheduler::notifyError(std::vector<folly::Promise<Status
...
 
+folly::Future<Status> AsyncMsgNotifyBasedScheduler::runExecute(Executor* executor) const {
+  // catch Executor::execute here, upward call stack should only get Status, no exceptions.
+  try {
+    addExecuting(executor);
+    if (hasFailStatus()) return failedStatus_.value();
+    folly::Future<Status> status = Status::OK();
+    {
+      memory::MemoryCheckGuard guard;
+      status = executor->execute();
```
Comment (yixinglu) [inline src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp:270]: plz NOTE: the `execute` is an async function, your memory check guard may be invalid when the tasks really run.

File: src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp [hunk 5]
```diff
@@ -225,17 +239,94 @@ void AsyncMsgNotifyBasedScheduler::notifyError(std::vector<folly::Promise<Status
...
+  return ss.str();
+}
+
+void AsyncMsgNotifyBasedScheduler::appendExecutor(size_t spaces,
+                                                  Executor* executor,
+                                                  std::stringstream& ss) {
+  ss << std::string(spaces, ' ') << formatPrettyId(executor) << std::endl;
+  for (auto depend : executor->depends()) {
+    appendExecutor(spaces + 1, depend, ss);
+  }
```
Comment (yixinglu) [inline src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp:328]: do the loop body executors need to be considered here?

File: src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp (hunk 4 above)
Comment (codesigner) [reply src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp:270]: > plz NOTE: the `execute` is an async function, your memory check guard may be invalid when the tasks really run.

This MemoryCheckGuard is only used by some Executor return a already satisfied future, like DedupExecutor::execute, Async future's MemoryCheckGuard is not here, they are declare inside, for example in StorageClientBase-inl.h:

```
  return folly::collectAll(respFutures)
      .deferValue([this, requests = std::move(requests), totalLatencies, hosts](
                      std::vector<folly::Try<StatusOr<Response>>>&& resps) {
        // throw in MemoryCheckGuard verified
        memory::MemoryCheckGuard guard;
        StorageRpcResponse<Response> rpcRes
```

File: src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp (hunk 5 above)
Comment (codesigner) [reply src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp:328]: > do the loop body executors need to be considered here?

no, just print a original dependency tree.

File: src/common/memory/MemoryUtils.cpp (hunk 1 above)
Comment (codesigner) [reply src/common/memory/MemoryUtils.cpp:173]: > best not to use equals to compare floating point data types

I had considered it as a issue, but I tried in real run environment, it works as expected.

**Review by yixinglu — APPROVED** [review]
LGTM

**Review by critical27 — APPROVED** [review]

**Pull request merged**